### ISSUE TITLE
Clarification on WCAG 2.4.6: Headings and Labels - Descriptiveness of accessible name for buttons (represented by icon)

### ISSUE THREAD

**Issue by Vaibosnikov** [w3c/wcag#4486]
Recently auditing a website where the button's (denoted by icon) accessible name in the DOM structure is provided with the use of `aria-label="button_name"` but it is not descriptive enough to orient the users with the context and purpose in the page. The WCAG [interpretation doc](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html) mentions that

> Further, in the case of labels, this success criterion does not take into consideration whether or not alternative methods of providing an accessible name for form controls and inputs have been used — that aspect is covered separately by [4.1.2: Name, Role and Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value). It is possible for controls and inputs to have an appropriate accessible name (e.g. using aria-label="…") and therefore pass Success Criterion 4.1.2, but to still fail this success criterion (if the label is inaccurate or insufficiently clear or descriptive).

Seeking clarification on the success criteria whether the descriptiveness of the label applies to visual label only or on accessible name as well.

**Comment by detlevhfischer** [comment]
> Seeking clarification on the success criteria whether the descriptiveness of the label applies to visual label only or on accessible name as well.

This is a well-spotted gap - in the absence of an SC covering descriptiveness of programmatic labels, we tend to report it as 2.4.6 issue (but there can be an argument that it should go to 4.1.2). Either way, it fails here or there.

**Comment by Vaibosnikov** [comment]
At first glance, the issue appears to fall under [WCAG 4.1.2: Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value), which normatively requires:

> Give components correct names, roles, states, and values.”

Here, the term "correct name" refers to ensuring that a programmatically determinable name is exposed to assistive technologies (AT) not necessarily whether that name is descriptive or contextually informative.

In the scenario described a button with `aria-label=" "` (an empty accessible name), despite having a visual label like "Go", fails 4.1.2 because the screen reader receives no accessible name at all. The semantic role may still be recognized (e.g., <button>), but the absence of a programmatically associated name renders the control unusable for screen reader users. Thus, it's a clear failure of 4.1.2.

If the button has a name like `aria-label="Go"` or `aria-label="Click"` which is technically exposed to AT but lacks descriptive clarity about the control’s purpose then it still passes 4.1.2, but fails WCAG 2.4.6: Headings and Labels.

That said, there remains some confusion: if a name is exposed but is so generic that it doesn't assist the user in understanding the purpose (e.g., aria-label="button"), then should this be more strongly aligned with 2.4.6 due to lack of descriptiveness, or can it still be interpreted as a 4.1.2 failure under the expectation of a "correct" name?

**Comment by detlevhfischer** [comment]
> That said, there remains some confusion: if a name is exposed but is so generic that it doesn't assist the user in understanding the purpose (e.g., aria-label="button"), then should this be more strongly aligned with 2.4.6 due to lack of descriptiveness, or can it still be interpreted as a 4.1.2 failure under the expectation of a "correct" name?

2.4.6 originally targeted visible labels and headings but since the issue (descriptiveness) is the same regardless whether a label is visible or just a programmatically available I think filing things like `aria-label="button"` under 2.4.6 is a good idea. But surely there will be others opting for filing it under 4.1.2.

**Comment by GreggVan** [comment]
isn't NAME what we call the invible programmatically determinable text for a control

and LABEL is the visible text with a control?\
- remembering that labels may or may not be programmatically determinable as associated with an element (e..g. if someone positions text near a control but it is not part of the control definition)

I see comments that seem to equate them (though they sometimes are the same text - they are different things and technically can have different content -- and somethimes should if the name contains additional visual information not needed in the label) 

Might be good to take care in our disscussions as to which we are talking about.

**Comment by detlevhfischer** [comment]
@GreggVan 
> if the name contains additional visual information not needed in the label

I guess you meant "if the label contains additional visual information not needed in the name"? 
E.g., a field with a visible label "Name (required)" but the programmatic name might be `aria-label="Name"` (with `required` set as input attribute, so dropping the (required) in the programmatic name to reduce redundancy? Or what scenario do you have in mind?

**Comment by GreggVan** [comment]
No I  meant if the Name needs additional information not needed in the label.   

A sighted person can see the label -- but also the page, its organization, and what is around it visually.     So less information may be needed in a label to understand it.

A blind person needs to rely more on the name.   So they may need more information.    

for example  lets say there is are several button that say   "OK"   but they are all next to different things so it is clear which accept button is for what.       The NAME for that button should not just be  "OK" but also include enough additional information to know what it applies to.

**Comment by patrickhlauke** [comment]
> lets say there is are several button that say "OK" but they are all next to different things so it is clear which accept button is for what. The NAME for that button should not just be "OK" but also include enough additional information to know what it applies to.

that would normally be handled under 1.3.1 Info and Relationships, no?
